arraysandpointers.tex

```diff
@@ -1,144 +1,166 @@
 
-\chapter{Arrays and Pointers}
+\chapter{Arrays and Memory Allocation}
 
-So far we have only talked about single value types, such as \texttt{int} and \texttt{char}.
-Arrays are also an important aspect to a programming language, and C has these as well.
+So far we have only talked about single-value types, such as \texttt{int} and \texttt{char}.
+The simplest multi-value type is an array, which stores multiple values of another type.
+For example, you could have an array of 8 \texttt{int}s.
 
 \section{Declaring an array}
 
 Given a type \texttt{T}, you declare an array of \texttt{T} using square brackets as follows:
-
 \begin{codeinline}
-T array[numElements];
+T arrayVariable[numElements];
 \end{codeinline}
-
-Where \texttt{numElements} is the number of elements you want in the array.
-Note than until C99, numElements had to be a compile time constant.
-This means that in general, if you want to declare an array you must already know the size of it.
-This is not always possible, and how to deal with that will be covered later in this section.
+where \texttt{numElements} is the number of elements you want in the array.
+Note that \texttt{numElements} must be a compile-time constant in C90, but can be a runtime value in C99.
+This means that you generally need to know the size of an array before you declare it, which is not always possible.
+This issue is dealt with later in this chapter.
 
 Below are some example array declarations.
 These also show how you can make arrays of arrays.
+Arrays can be nested arbitrarily to create multidimensional arrays of any dimension.
 
 \begin{codeblock}
 int nums[6]; // Array of 6 ints
-int 2dNums[3][4]; // Array of 3 arrays of 4 ints
+int twoDimenionsOfNums[3][4]; // Array of 3 arrays of 4 ints
+\end{codeblock}
+
+\section{Initializing an array}
+
+When you declare a variable (including arrays) in a function, C allocates memory for it but does not initialize the memory to any particular value.
+This means that if we just declare an array like we did above, it will contain gibberish elements.
+We could loop through the array and assign each element to something sensible, or we could use C's array initialization syntax:
+
+\begin{codeblock}
+int a[4] = { 0, 1, 5, 6 }; // All elements are initialized
+int b[4] = { 4, 3 }; // Remaining elements are initialized to 0
+int c[4] = { }; // All elements are initialized to 0
```

Comment:
test.c:29:16: warning: ISO C forbids empty initializer braces [-Wpedantic]
     int c[4] = {};

You always have to initialise 1 element. Usually you would do something like

int c[4] = { 0 };